File: src/main/java/me/meta4245/betterthanmodern/event/ItemRegistry.java
package me.meta4245.betterthanmodern.event;

import net.mine_diver.unsafeevents.listener.EventListener;
import net.minecraft.item.Item;
import net.modificationstation.stationapi.api.event.registry.ItemRegistryEvent;
import net.modificationstation.stationapi.api.mod.entrypoint.Entrypoint;
import net.modificationstation.stationapi.api.util.Identifier;
import net.modificationstation.stationapi.api.util.Namespace;
import net.modificationstation.stationapi.api.util.Null;

import java.lang.reflect.Field;

import static me.meta4245.betterthanmodern.reflection.Utilities.*;

public class ItemRegistry {
    @Entrypoint.Namespace
    public static Namespace NAMESPACE = Null.get();

    public static Item blocksDisc;
    public static Item carrot;
    public static Item catDisc;
    public static Item chirpDisc;
    public static Item elevenDisc;
    public static Item farDisc;
    public static Item glisteringMelon;
    public static Item goldenCarrot;
    public static Item goldNugget;
    public static Item mallDisc;
    public static Item mellohiDisc;
    public static Item melonSeeds;
    public static Item stalDisc;
    public static Item stradDisc;
    public static Item thirteenDisc;
    public static Item waitDisc;
    public static Item wardDisc;
    public static Item cookedChicken;
    public static Item cookedMutton;
    public static Item cookedPorkchop;
    public static Item steak;
    public static Item rawBeef;
    public static Item rawChicken;
    public static Item rawMutton;
    public static Item rawPorkchop;
    public static Item melonSlice;
    public static Item potato;
    public static Item bakedPotato;

    private Item item(Class<? extends Item> clazz) {
        String key = namespaceName(clazz);
        Item item;

        try {
            item = clazz.getConstructor(Identifier.class)
                    .newInstance(NAMESPACE.id(key))
                    .setTranslationKey(NAMESPACE, key);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }

        return item;
    }

    @EventListener
    public void registerItems(ItemRegistryEvent event) {
        getItems().forEach(i -> {
            Field f = getField(ItemRegistry.class, fieldName(i));
            setField(f, item(i));
        });
    }
}

File: src/main/kotlin/me/meta4245/betterthanmodern/event/ItemRegistry.kt
package me.meta4245.betterthanmodern.event

import net.minecraft.item.Item
import net.modificationstation.stationapi.api.mod.entrypoint.Entrypoint
import net.modificationstation.stationapi.api.util.Identifier
import net.modificationstation.stationapi.api.util.Namespace

class ItemRegistry {
    companion object {
        @Entrypoint.Namespace
        @JvmStatic
        val NAMESPACE: Namespace? = null
    }

    private fun item(clazz: Class<out Item>): Item {
        val name = StringBuilder()
        val className = clazz.simpleName
        val len = className.length

        for (i in 0 until len) {
            val c = className[i]
            if (Character.isUpperCase(c) && i != len - 1 && i != 0) {
                name.append("_")
            }
            name.append(c.lowercaseChar())
        }

        val key = name.toString()
        val i: Item

        try {
            i = clazz.getConstructor(Identifier::class.java).newInstance(
                NAMESPACE?.id(key)
            ).setTranslationKey(NAMESPACE, key)
        } catch (e: Exception) {
            throw RuntimeException(e)
        }

        return i
    }
}
